pkg/proto/stream -- streaming PB encoding

djmitche opened the pull request:
### What does this PR do?

This introduces a new Go package that supports "streaming" protobuf.  See `docs.go` for details.

### Motivation

The marshaling-based PB support is pretty slow and allocation-heavy for the purposes we put it to in creating payloads for the intakes.  Particularly since we must break objects into pieces to fit into limited-size payloads.

The sketch serialization currently uses a bit of hand-rolled serialization to get good performance, but it's "magic".

This library provides a less-magic way to accomplish that hand-rolled serialization, and more.

A test port of the sketch serialization to use this library shows about 5% speed increase and 50% reduction in allocations.

### Describe how to test your changes

Run the unit tests.  This PR does not use the new functionality (that will come later).

### Checklist

- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.

djmitche committed:
pkg/proto/stream -- streaming protobuf encoding

djmitche committed:
fix lint

djmitche committed:
avoid alloc in ProtoStream#String

sgnn7 reviewed (commented):
@djmitche Left some comments - mostly nits and stylistic stuff

djmitche commented on pkg/proto/stream/example_test.go:
```diff
@@ -0,0 +1,237 @@
+
+	err = ps.String(fieldQuery, "q=streaming+protobufs")
+	if err != nil {
+		panic(err)
```
Examples can't return errors, it seems:
```
.../_testmain.go:45:46: cannot use stream.ExampleProtoStream_EmbeddedMessage (type func() error) as type func() in field value
```

sgnn7 replied:
Yeah definitely missed that you were using the doc examples for this 👍

djmitche commented on pkg/proto/stream/example_test.go:
```diff
@@ -0,0 +1,237 @@
+	"github.com/golang/protobuf/proto"
+)
+
+func ExampleNew() {
```
The pattern for examples is `ExampleF` where `F` is the function for which it is an example.

djmitche commented on pkg/proto/stream/example_test.go:
```diff
@@ -0,0 +1,237 @@
+	// query 1: author=rumi
+}
+
+func ExampleProtoStream_Sint32Packed() {
```
The `_test.go` means it's only built for tests; and the `Example` functions [are tests](https://pkg.go.dev/testing#hdr-Examples).

djmitche commented on pkg/proto/stream/stream.go:
```diff
@@ -0,0 +1,548 @@
+	child *ProtoStream
+
+	// childBuffer is the buffer to which `child` writes.
+	childBuffer bytes.Buffer
```
The second one relates to the first -- it's the buffer to which `child` writes.  So I'm not sure how I could change the naming?

sgnn7 replied:
I think just suffixing these would make the members as a whole a bit more consistent:
- `output` -> `outputWriter`
- `scratch` -> `scratchBuffer`

djmitche commented on pkg/proto/stream/stream_test.go:
```diff
@@ -0,0 +1,286 @@
+	t.Run("DoublePacked", func(t *testing.T) {
+		output.Reset()
+		require.NoError(t, ps.DoublePacked(fieldDub, []float64{3.14, 1.414}))
+		assertBytes(t, []byte{key(fieldDub), 0x10, 0x1f, 0x85, 0xeb, 0x51, 0xb8, 0x1e, 0x9, 0x40, 0x39, 0xb4, 0xc8, 0x76, 0xbe, 0x9f, 0xf6, 0x3f}, output.Bytes())
```
Yeah, I can break that down with some comments as to what each byte means.

Protobufs are specified independent of a library or version, so there's no need to reference a specific library + version.  If we needed to update this, that would due to an incompatible change in proto (old and new payloads would not be compatible -- something protobufs exists to avoid).

djmitche replied:
Encoding is documented [here](https://developers.google.com/protocol-buffers/docs/encoding)

djmitche commented on pkg/proto/stream/example_test.go:
```diff
@@ -0,0 +1,241 @@
+import (
+	"bytes"
+	"encoding/binary"
+	fmt "fmt"
```
goimport bug??

djmitche commented on pkg/proto/stream/test.proto:
```diff
@@ -0,0 +1,43 @@
+// This is only used for tests in this package.  Regenerate with
```
I don't want to keep up with upgrades -- this is only for testing, and only to verify that the encoded data can be decoded by a known-good implementation.  So, no need to continue to upgrade that implementation.

djmitche commented on pkg/proto/stream/stream.go:
```diff
@@ -0,0 +1,548 @@
+// will write their output to the wrapped `io.Writer`.  Zero values are not included.
+//
+// ProtoStream instances are *not* threadsafe and *not* re-entrant.
+type ProtoStream struct {
```
An interface would introduce a dynamic dispatch (indirect jump).  Since this is used in hot loops, that is potentially a high cost.  Types with no public [are](https://pkg.go.dev/bytes#Buffer) [pretty](https://pkg.go.dev/math/rand#Rand) [common](https://pkg.go.dev/os#File).

sgnn7 replied:
Yeah fair points

djmitche commented on pkg/proto/stream/stream_test.go:
```diff
@@ -0,0 +1,286 @@
+const fieldBoo = 13
+const fieldStr = 14
+const fieldByt = 15
+const fieldEnu = 16
```
(admittedly these are for tests, but I'd like to set a good example)

sgnn7 replied:
> I don't think that'd be a good idea, since these correspond to the values in the .proto file, which are specified explicitly

I see your point though some kind of a comment to indicate this would be nice on top of the `const` stack.

djmitche committed:
address @sgnn7's review

djmitche committed:
add a required Reset method

djmitche committed:
update for reviews from @richardartoul

djmitche committed:
Avoid a copy for packed encodings

This shows about a 10% speed increase, without any additional
allocations.

djmitche committed:
Switch to using protowire directly

The `proto.Buffer` struct just calls through to protowire.  It's also
deprecated.

This change brings a ~15% speed increase for simple encoding operations.

djmitche commented:
I think I've addressed all of the review comments.  I also optimized the packed encoding a bit, and switched from `proto.Buffer` to `protowire.Append..`.

gh123man reviewed (approved):
👍

sgnn7 reviewed (approved):
👍

djmitche commented:
(all of the CI failures look like the usual intermittents)

truthbk commented on go.mod:
```diff
@@ -185,6 +185,7 @@ require (
 	gomodules.xyz/jsonpatch/v3 v3.0.1
 	google.golang.org/genproto v0.0.0-20210602131652-f16073e35f0c
 	google.golang.org/grpc v1.38.0
+	google.golang.org/protobuf v1.26.0
```
I am not sure if we can make this leap, if I recall correctly we already have some pins to older versions that were required by grpc, and in turn, those were forced by an etcd library (`go.etcd.io/etcd/client/v2`).

djmitche replied:
The updated version of this package, in Molecule, has no (non-test) dependencies, so this should be fine.

truthbk commented on pkg/proto/stream/test.proto:
```diff
@@ -0,0 +1,43 @@
+// This is only used for tests in this package.  Regenerate with
+//
+//  (cd pkg/proto/stream; protoc --go_out=. --go_opt=paths=source_relative test.proto && mv test.pb.go stream_pb_test.go)
```
We have some invoke tasks for protobuf generation; I understand this is for testing, but I think these commands should be added there, see: https://github.com/DataDog/datadog-agent/blob/main/tasks/go.py#L457

djmitche replied:
This got moved out to Molecule, too, and I fitted this into the pre-existing generated code for testing.

truthbk reviewed (commented):
Couple of minor comments, nice stuff! 👍

ogaca-dd reviewed (commented):
Great pull request!

djmitche commented on pkg/proto/stream/stream.go:
```diff
@@ -0,0 +1,510 @@
+//
+// NOTE: if the inner function creates an empty message (such as for a struct
+// at its zero value), that empty message will still be added to the stream.
+func (ps *ProtoStream) Embedded(fieldNumber int, inner func(*ProtoStream) error) error {
```
Yes, you'd just have nested function calls.

```go
err = ps.Embedded(fieldSubMessage, func(ps *ProtoStream) error {
    err = ps.Embedded(fieldSubSubMessage, func(ps *ProtoStream) error {
        err = ps.String(fieldSubSubMessageFo, "bar")
        if err != nil {
            return err
        }
    })
    if err != nil {
        panic(err)
    }
})
if err != nil {
    panic(err)
}
```

djmitche commented on pkg/proto/stream/stream.go:
```diff
@@ -0,0 +1,510 @@
+	BufferFactory func() []byte
+}
+
+// NewProtoStream creates a new ProtoStream.  This ProtoStream _cannot be used_ until
```
I tend to agree here.  As a compromise, I'll add the writer back to thew NewMethod, but keep the reset as well.

djmitche commented on pkg/proto/stream/stream.go:
```diff
@@ -0,0 +1,510 @@
+}
+
+// Int64 writes a value of proto type int64 to the stream.
+func (ps *ProtoStream) Int64(fieldNumber int, value int64) error {
```
This relies on the compiler to inline one function into the other, or pay the cost of an additional function call.  Is that worthwhile?

djmitche replied:
I used to have a `resetScratch` method that would reset the scratch buffer, and noticed a bump in performance by replacing that function call with the function body.  So I'm hesitant to rely too much on the compiler's inlining skills.

ogaca-dd replied:
It seems Go compiler does not inline the function in this case. Do you have an idea of the cost of serializing a whole object when there is no inlining?

djmitche replied:
I didn't look at the source to see what's inlined and what isn't, or why, or how to change that, so I don't have any comparative benchmarks.

I think this change would reduce clarity fractionally -- the regularity of these methods is, I think, pretty readable and not excessively verbose.

djmitche commented:
I've made the change to NewProtoStream [here](https://github.com/richardartoul/molecule/pull/9).  Once the other comments here are resolved, I'll close this PR in favor of that one.

djmitche commented:
I've addressed the comments here, and made the relevant adjustments in the molecule PR mentioned above.  Thanks!
